### ISSUE TITLE
Access components of specific Entity

### ISSUE THREAD

**Issue by itmuckel** [slide-rs/specs#334]
I created an entity with a "Visible(bool)" component that is false by default.
In a specific System I need to set this component to true.

So basically I need to get a specific Entity and set it's component. Can I do this in a System? Should the Entity be a resource? How does the "Entities" Storage work? I couldn't find answers in closed issues or the docs.

The naive approach would be to give this entity a component that's unique to it (IsTheComponentImLookingFor) and query that in a system, but this seems sooo wrong. 😆

**Comment by WaDelma** [comment]
You could make the entity resource if there is truly only one of them and then get it's component:
```rust
struct SpecialEntity(Entity);
// When initialising game: world.add_resource(SpecialEntity(entity));

struct Sys;
impl<'a> System<'a> for Sys {
    type SystemData = (WriteStorage<'a, Visible>, Fetch<'a, SpecialEntity>);
    fn run(&mut self, (mut visible, special_entity): Self::SystemData) {
        *visible.get_mut(special_entity.0)
            .expect("In this example there is no error handling")
            .0 = true;
    }
}
```
But I would prefer giving the entity marker component so that eventually there can be multiple of them without breaking assumptions:
```rust
#[derive(Component, Default)]
#[component(NullStorage)]
struct SpecialEntity;

struct Sys;
impl<'a> System<'a> for Sys {
    type SystemData = (WriteStorage<'a, Visible>, ReadStorage<'a, SpecialEntity>);
    fn run(&mut self, (mut visible, special_entity): Self::SystemData) {
        for (visible, _) in (&mut visible, &special_entity.check()).join() {
             visible.0 = true;
        }
    }
}
```

**Comment by torkleyy** [comment]
I fully agree with @WaDelma (thank you also for the example code) and I want to add a third possibility: You can also store the entity in the system if you only need to acces it from one system. However, generall option 2 (a marker component) is preferable.

**Comment by itmuckel** [comment]
Okay, your suggestions gave me some ideas for experimentation, but my problem is a bit more complex.
The concrete example is this. I want to show a destination mark, when the player has a current destination. The code I have is this:

```rust
impl<'a> System<'a> for MoveToDestination {
    type SystemData = (WriteStorage<'a, CurrentDestination>,
                       WriteStorage<'a, Position>,
                       WriteStorage<'a, Visible>,
                       ReadStorage<'a, DestinationMarkMarker>);

    fn run(&mut self, data: Self::SystemData) {
        let (mut dest, mut pos, mut visible, destination_mark_marker) = data;

        for (mut dest, mut pos) in (&mut dest, &mut pos).join() {
            if let Some(destination) = dest.0 {
                let (new_pos,
                    new_dest) = move_to_point(&Point::new(pos.x, pos.y),
                                        &Some(destination));

                *dest = CurrentDestination(new_dest);
                pos.x = new_pos.x;
                pos.y = new_pos.y;

                // set position and visibility of destination marker.
            } else {
                // set visibility of destination marker to false.
            }
        }
    }
}
```

How would I get the entity with the destination marker at the specified locations *and* set its components?

**Comment by WaDelma** [comment]
I would have component `CurrentDestination` that contains `Entity` which references the destination marker entity which has position. When the destination is reached `CurrentDestination` and the destination marker entity will both be removed.

**Comment by itmuckel** [comment]
Thank you for your quick response, but I don't understand. Could you add some code? 

Is it generally not advisable, what I'm trying to do there? I still think a lot like "If this happens, then this entity should do that. If this entity does this, then that entity should do something else...". I'm still new to the way entities and systems work together and especially how to structure systems that somehow depend on each other.

**Comment by itmuckel** [comment]
Okay, I've got something working. I can't say I like it, but it works for the moment.

I left the system for handling movement towards a destination as it is.<br>
I created a system specifically for handling the destination marker.

First it queries the current destination. If it is set, it saves its destination in a local variable.<br>
Then it queries the destination marker (tagged by a special component, like you two suggested). It sets the destination marker's properties accordingly. 

```rust
pub struct HandleDestinationMarker;

impl<'a> System<'a> for HandleDestinationMarker {
    type SystemData = (WriteStorage<'a, Position>,
                       WriteStorage<'a, Visible>,
                       ReadStorage<'a, CurrentDestination>,
                       ReadStorage<'a, DestinationMarkMarker>);

    fn run(&mut self, data: Self::SystemData) {
        let (mut pos, mut visible, current_dest, dest_marker) = data;

        let mut draw_dest_at_point: Option<Point> = None;

        for current_dest in (&current_dest).join() {
            if let Some(dest) = current_dest.0 {
                draw_dest_at_point = Some(Point::new(dest.x, dest.y))
            }
        }

        for (mut visible, mut pos, _) in (&mut visible, &mut pos, &dest_marker).join() {
            if let Some(point) = draw_dest_at_point {
                pos.x = point.x;
                pos.y = point.y;
                visible.0 = true;
            } else {
                visible.0 = false;
            }
        }
    }
}
```

**Comment by WaDelma** [comment]
```rust
#[derive(Component, Clone, Copy)]
struct CurrentDestination(Entity);

#[derive(Component, Clone, Copy)]
struct Position(f32, f32);


struct Sys;
impl<'a> System<'a> for Sys {
    type SystemData = (
        Entities<'a>,
        WriteStorage<'a, Position>,
        WriteStorage<'a, CurrentDestination>
    );

    fn run(&mut self, (entities, mut positions, mut destinations): Self::SystemData) {
        let to_be_moved = (&*entities, &destinations)
            .join()
            .map(|(a, &CurrentDestination(b))| (a, b))
            .collect::<Vec<_>>();

        for (entity, dest) in to_be_moved {
            if let (Some(destination), Some(position)) = (positions.get(dest).cloned(), positions.get_mut(entity)) {
                if are_we_there(position, &destination) {
                    entities.delete(dest).unwrap();
                    destinations.remove(entity);
                } else {
                    move_it(position, &destination);
                }
            }
        }
    }
}
```

This is what I was thinking, but apparently you got it working while I cooked that up. :D

**Comment by itmuckel** [comment]
@WaDelma 
Doesn't matter, it's always good to see more than one solution to a problem. For example I didn't know you could use `Entities` like that. Thanks for your contribution. 😄

**Comment by torkleyy** [comment]
@WaDelma Do you feel there is anything we should document here?

**Comment by WaDelma** [comment]
The book should probably have chapter dedicated for components that would explain stuff like storing entities in components and using marker components vs "special" entities.

**Comment by Rhuagh** [comment]
Fixed by #395.
